### PR TITLE
fix(be): Modify the upstream to use the new JSON patch package

### PR DESCRIPTION
**Pull request by Jaycean**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues
#1392 
#1332
___
### Bugfix
- Description
I find that the jsonpatch(github.com/api7/go-jsonpatch) used by upstream and service when calling the patch method is still the old package,updated to github.com/evanphx/json-patch/v5

- How to fix?
Modify the upstream to use the new JSON patch package

### CODE REVIEW

**Comment by imjoey** [comment]
@Jaycean In order to be a semantic pull request, please change the title to `fix: Modify the upstream to use the new JSON patch package` . Thanks.

**Review by membphis — APPROVED** [review]

File: api/internal/handler/upstream/upstream.go [hunk 1]
```diff
@@ -17,11 +17,11 @@
 package upstream
 
 import (
+	"encoding/json"
 	"net/http"
 	"reflect"
 	"strings"
 
-	"github.com/api7/go-jsonpatch"
```
Comment (nic-chen) [inline api/internal/handler/upstream/upstream.go:24]: seems we don't need `go-jsonpatch` any more. try `go mod tidy` to see if we could remove it？

**Review by imjoey — CHANGES_REQUESTED** [review]
@Jaycean  Since #1396 has been merged already, the Upstream is the only resource that's using `github.com/api7/go-jsonpatch`. So within this PR, we would better completely remove the dependency of `github.com/api7/go-jsonpatch`. 

Please rebase the master branch with the latest code and run `go mod tidy` for that. Thanks.

**Comment by Jaycean** [comment]
> @Jaycean Since #1396 has been merged already, the Upstream is the only resource that's using `github.com/api7/go-jsonpatch`. So within this PR, we would better completely remove the dependency of `github.com/api7/go-jsonpatch`.
> 
> Please rebase the master branch with the latest code and run `go mod tidy` for that. Thanks.

Thanks for the reminder.done.

**Comment by imjoey** [comment]
@nic-chen Backend E2E Test seems broken since this morning, prompted with the following error.  Could you please have a check? Thanks.

```
--- FAIL: TestBalancer_roundrobin_with_weight (0.66s)
    printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/upstreams/1
    --- PASS: TestBalancer_roundrobin_with_weight/create_upstream_(roundrobin_with_same_weight) (0.06s)
    printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/routes/1
    --- PASS: TestBalancer_roundrobin_with_weight/create_route_using_the_upstream_just_created (0.31s)
    base.go:150: 
        	Error Trace:	base.go:150
        	            				balancer_test.go:77
        	Error:      	Expected nil, but got: &url.Error{Op:"Get", URL:"http://127.0.0.1:9081/server_port", Err:(*net.OpError)(0xc0001b61e0)}
        	Test:       	TestBalancer_roundrobin_with_weight
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x8c8eed]
```

File: api/internal/handler/upstream/upstream.go [hunk 2]
```diff
@@ -198,39 +201,36 @@ func (h *Handler) BatchDelete(c droplet.Context) (interface{}, error) {
...
-	input := c.Input().(*UpdateInput)
-	arr := strings.Split(input.ID, "/")
-	var subPath string
-	if len(arr) > 1 {
-		input.ID = arr[0]
-		subPath = arr[1]
-	}
+	input := c.Input().(*PatchInput)
+	reqBody := input.Body
+	ID := input.ID
```
Comment (tokers) [inline api/internal/handler/upstream/upstream.go:213]: we should use variables started with letter in small case.

File: api/internal/handler/upstream/upstream.go [hunk 3]
```diff
@@ -198,39 +201,36 @@ func (h *Handler) BatchDelete(c droplet.Context) (interface{}, error) {
...
+	res, err := utils.MergePatch(stored, subPath, reqBody)
+
+	if err != nil {
+		return handler.SpecCodeResponse(err), err
 	}
 
-	if err := patch.Apply(&stored); err != nil {
+	var upstream entity.Upstream
+	err = json.Unmarshal(res, &upstream)
+	if err != nil {
```
Comment (tokers) [inline api/internal/handler/upstream/upstream.go:229]: Can be merged into:

```go
if err := json.Unmarshal(res, &upstream); err != nil {
```

Comment (tokers) [inline api/internal/handler/upstream/upstream.go:229]: Can be merged into:

```go
if err := json.Unmarshal(res, &upstream); err != nil {
```

File: api/internal/handler/upstream/upstream.go [hunk 4]
```diff
@@ -314,3 +314,4 @@ func (h *Handler) listUpstreamNames(c droplet.Context) (interface{}, error) {
 
 	return output, nil
 }
+
```
Comment (tokers) [inline api/internal/handler/upstream/upstream.go:317]: Why keep an empty line here?

**Comment by nic-chen** [comment]
> @nic-chen Backend E2E Test seems broken since this morning, prompted with the following error. Could you please have a check? Thanks.
> 
> ```
> --- FAIL: TestBalancer_roundrobin_with_weight (0.66s)
>     printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/upstreams/1
>     --- PASS: TestBalancer_roundrobin_with_weight/create_upstream_(roundrobin_with_same_weight) (0.06s)
>     printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/routes/1
>     --- PASS: TestBalancer_roundrobin_with_weight/create_route_using_the_upstream_just_created (0.31s)
>     base.go:150: 
>         	Error Trace:	base.go:150
>         	            				balancer_test.go:77
>         	Error:      	Expected nil, but got: &url.Error{Op:"Get", URL:"http://127.0.0.1:9081/server_port", Err:(*net.OpError)(0xc0001b61e0)}
>         	Test:       	TestBalancer_roundrobin_with_weight
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> 	panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x8c8eed]
> ```

@imjoey  @Jaycean  hi, have you found the reason of the failure of ci?

**Comment by Jaycean** [comment]
> > @nic-chen Backend E2E Test seems broken since this morning, prompted with the following error. Could you please have a check? Thanks.
> > ```
> > --- FAIL: TestBalancer_roundrobin_with_weight (0.66s)
> >     printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/upstreams/1
> >     --- PASS: TestBalancer_roundrobin_with_weight/create_upstream_(roundrobin_with_same_weight) (0.06s)
> >     printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/routes/1
> >     --- PASS: TestBalancer_roundrobin_with_weight/create_route_using_the_upstream_just_created (0.31s)
> >     base.go:150: 
> >         	Error Trace:	base.go:150
> >         	            				balancer_test.go:77
> >         	Error:      	Expected nil, but got: &url.Error{Op:"Get", URL:"http://127.0.0.1:9081/server_port", Err:(*net.OpError)(0xc0001b61e0)}
> >         	Test:       	TestBalancer_roundrobin_with_weight
> > panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> > 	panic: runtime error: invalid memory address or nil pointer dereference
> > [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x8c8eed]
> > ```
> 
> @imjoey @Jaycean hi, have you found the reason of the failure of ci?
At present, we can only locate there
```
func BatchTestServerPort(t *testing.T, times int) map[string]int {
	url := APISIXSingleWorkerHost + "/server_port"
	req, err := http.NewRequest(http.MethodGet, url, nil)
	assert.Nil(t, err)

APISIXSingleWorkerHost = "http://127.0.0.1:9081"
```
This request`http://127.0.0.1:9081/server_port` 
A new route is created, which is exposed by apimix. It looks like a network problem here, but it can't be reproduced locally. No further problem is found.

**Comment by imjoey** [comment]
> @imjoey @Jaycean hi, have you found the reason of the failure of ci?

@nic-chen  sorry, not get the solution yet. 😢

**Comment by nic-chen** [comment]
@Jaycean @imjoey  
It seems to be affected by the last merged commit of APISIX

**Comment by Jaycean** [comment]
> @Jaycean @imjoey
> It seems to be affected by the last merged commit of APISIX

Yes, just now I pulled the latest code of apisix, and the error has been reproduced locally.

**Comment by nic-chen** [comment]
@Jaycean  CI passed

**Comment by imjoey** [comment]
> @Jaycean CI passed

@nic-chen great thanks.

File: api/internal/handler/upstream/upstream.go (hunk 3 above)
Comment (Jaycean) [reply api/internal/handler/upstream/upstream.go]: done. Thks

**Review by imjoey — APPROVED** [review]
/lgtm

Thanks @Jaycean .

**Comment by Jaycean** [comment]
> @Jaycean CI passed

Great Thks.

**Pull request merged**